### PR TITLE
Tf-idf

### PR DESCRIPTION
**Pull request by mat-shyR**
Created new text.py file containing tfidf class.

### CODE REVIEW

File: verticapy/machine_learning/feature_extraction/text.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+from verticapy._typing import SQLRelation
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:1]: please add OpenText header string. It is in all our files. Also format your code using the latest version of black.

File: verticapy/machine_learning/feature_extraction/text.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        processing.
+    """
+    def __init__(
+        self,
+        name: str = None,
+        schema: str = None,
+        overwrite_model: bool = False,
+        lowercase: bool = True
+    ) -> None:
+         self.parameters = {
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:30]: For the name, you should take it out from the parameters and use the super method:
```
super().__init__(name, overwrite_model)
```
The name includes the schema so no need for a schema parameter.

File: verticapy/machine_learning/feature_extraction/text.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        Name of the model.
+    schema: str
+        Name of the model.
+    overwrite_model: bool, default=False
+        If set to True, training a model with the same
+        name as an existing model overwrites the
+        existing model.
+    lowercase: bool, default=True
+        Converts  all  the elements to lowercase  before
+        processing.
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:21]: please ass an RST example: @mail4umar can help you.

File: verticapy/machine_learning/feature_extraction/text.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+             "name": name,
+             "schema": schema,
+             "overwrite_model": overwrite_model,
+             "lowercase": lowercase
+         }
+
+    def fit(
+        self,
+        input_relation: SQLRelation,
+        index: str,
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:40]: why not renaming column to x so we unify the fit functions (or X if you consider the possibility to merge multiple columns together in your function).

File: verticapy/machine_learning/feature_extraction/text.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+             "lowercase": lowercase
+         }
+
+    def fit(
+        self,
+        input_relation: SQLRelation,
+        index: str, 
+        column: str 
+        )-> None:
+        """Apply basic pre-processing. Create table with fitted vocabulary and idf values
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:43]: add the description in the next line

File: verticapy/machine_learning/feature_extraction/text.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        
+        Parameters
+        ----------
+        input_relation: SQLRelation (db object or vdf)
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:47]: no description?

File: verticapy/machine_learning/feature_extraction/text.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        
+        if not self.parameters["lowercase"]:
+            text = {column}
+        else:
+            text = f"LOWER({column})"
+        
+        self.idf_ = ".".join([self.parameters["schema"],self.parameters["name"]])
+
+        if self.parameters["overwrite_model"]:
+            _executeSQL(f"DROP TABLE IF EXISTS {self.idf_}", print_time_sql=False)
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:71]: why not using the drop function?

File: verticapy/machine_learning/feature_extraction/text.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            vdf = input_relation.copy()
+        else:
+            vdf = vDataFrame(input_relation)
+        
+        if not self.parameters["lowercase"]:
+            text = {column}
+        else:
+            text = f"LOWER({column})"
+        
+        self.idf_ = ".".join([self.parameters["schema"],self.parameters["name"]])
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:68]: use self.model_name

File: verticapy/machine_learning/feature_extraction/text.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+        if self.parameters["overwrite_model"]:
+            _executeSQL(f"DROP TABLE IF EXISTS {self.idf_}", print_time_sql=False)
+
+        q_idf = f"""CREATE TABLE {self.idf_} AS
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:73]: do you really need to create a table? Should it not be another method.
You should probably save your sql somewhere:
```
self.relation = ...
```
and use it later to build your vdf.

File: verticapy/machine_learning/feature_extraction/text.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+    def transform(
+        self,
+        input_relation: SQLRelation,
+        index: str,
+        column: str,
+        pivot: bool = False 
+        )-> vDataFrame:
+
+        """Apply basic pre-processing. Create table with vocabulary and idf values
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:125]: same as above

File: verticapy/machine_learning/feature_extraction/text.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        
+        Parameters
+        ----------
+        input_relation: SQLRelation (db object or vdf)
```
Comment (oualib) [inline verticapy/machine_learning/feature_extraction/text.py:129]: same as above

**Review by oualib — CHANGES_REQUESTED** [review]
Looks excellent:
Some things to address and we will be soon able to merge it.
Also, you should create an __init__ file for feature_extraction folder.

**Comment by mat-shyR** [comment]
Implemented same formula as Scikit-learn.

To do (will update):

- Add parameter to control normalization (at the moment only L2)
- Add parameter for smoothing (at the moment is always applied)
- Rewrite part of the queries to make them more readable
- ...

File: verticapy/machine_learning/feature_extraction/text.py (hunk 9 above)
Comment (mat-shyR) [reply verticapy/machine_learning/feature_extraction/text.py:73]: From my basic testing, creating a table makes the function much faster (4x).
Maybe I could add a parameter to let the user decide if he wants to save the model (pretrained dictionary and idf scores).

Comment (oualib) [reply verticapy/machine_learning/feature_extraction/text.py:73]: please give the option, it can be interesting

**Comment by oualib** [comment]
@mat-shyR I did all the changes needed to do a first merge: https://github.com/vertica/VerticaPy/pull/842/commits/f09b17cf84742ba35e41f671663fd3cda2beb339

Main TODO:
 - improve code quality.
 - Use RST to build up an example.
 - Unify variable names.
 - add other functions.

Other TODO
 - Add parameter to control normalization (at the moment only L2)
 - Add parameter for smoothing (at the moment is always applied)
 - Rewrite part of the queries to make them more readable

Please create a new PR for those changes and be sure to pull everything from master first.

**Review by oualib — APPROVED** [review]
